src/main/java/net/minecraftforge/client/model/generators/MultiPartBlockStateBuilder.java

```diff
@@ -107,30 +113,34 @@ public PartBuilder useOr() {
             Preconditions.checkArgument(values.length > 0, "Value list must not be empty");
             Preconditions.checkArgument(!conditions.containsKey(prop), "Cannot set condition for property \"%s\" more than once", prop.getName());
             Preconditions.checkArgument(canApplyTo(owner), "IProperty %s is not valid for the block %s", prop, owner);
+            Preconditions.checkState(nestedConditionGroups.isEmpty(), "Can't have normal conditions if there are already nested condition groups");
             this.conditions.putAll(prop, Arrays.asList(values));
             return this;
         }
 
+        /**
+         * Allows having nested groups of conditions if the current condition group is empty.
+         * @throws IllegalStateException if there are any normal conditions in the current condition group
+         */
+        public final ConditionGroup nestedGroup()
+        {
+            Preconditions.checkState(conditions.isEmpty(), "Can't have nested condition groups if there are already normal conditions");
+            ConditionGroup group = new ConditionGroup();
+            this.nestedConditionGroups.add(group);
+            return group;
+        }
+
         public MultiPartBlockStateBuilder end() { return MultiPartBlockStateBuilder.this; }
 
         JsonObject toJson() {
             JsonObject out = new JsonObject();
             if (!conditions.isEmpty()) {
-                JsonObject when = new JsonObject();
-                for (Entry<Property<?>, Collection<Comparable<?>>> e : conditions.asMap().entrySet()) {
-                    StringBuilder activeString = new StringBuilder();
-                    for (Comparable<?> val : e.getValue()) {
-                        if (activeString.length() > 0)
-                            activeString.append("|");
-                        activeString.append(((Property) e.getKey()).getName(val));
-                    }
-                    when.addProperty(e.getKey().getName(), activeString.toString());
-                }
-                if (useOr) {
-                    JsonObject innerWhen = when;
-                    when = new JsonObject();
-                    when.add("OR", innerWhen);
-                }
+                JsonObject when = MultiPartBlockStateBuilder.toJson(this.conditions, this.useOr);
+                out.add("when", when);
```

Comment:
This can just be inlined to `out.add("when", MultiPartBlockStateBuilder....` same with the one in the nestedConditionsGroup branch.

Reply:
Ah, I didn't catch that.